{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "93155487-49fa-4c6f-a4eb-5324bbfac1b7",
   "metadata": {},
   "source": [
    "# Prototype for transforming mutations into neoantigens"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9e22cf0-bcfe-43d1-b5da-31ab42e4bf60",
   "metadata": {},
   "source": [
    "## Input/output and requirements\n",
    "\n",
    "Our **input data** is going to come in three files:\n",
    "* VCF containing mutations\n",
    "* FASTA file containing the reference genome\n",
    "* GFF file containing the gene annotations\n",
    "\n",
    "The **expected output** is a neoantigen derived from a mutation in a given transcript.\n",
    "\n",
    "This prototype clears out the technicalities of manipulating the data to build neoantigen peptides for a list of mutations. For clarity purposes, some points are not considered yet:\n",
    "* Indels. Some details will need to be adapted for indels\n",
    "* Phasing of mutations. When two variants are in proximity we need to know if they occur in the same molecule or not, in order to build one or two neoantigen candidates. This information will be potentially provided in the VCF.\n",
    "* Transcripts in the negative strand. Reverse complement will need to be applied before translation.\n",
    "\n",
    "Some open questions remain such as:\n",
    "* Transcript selection. Which transcripts are we going to consider? Here we perform a sample filtering of 'protein_coding' genes that are flagged as 'basic'.\n",
    "* Indels overlapping intron and exon. What do we do with these? The simple alternative is to assume these do not disrupt the splicing.\n"
    "* Adjust neoantigens for germline mutations."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f17638b-742c-4b43-afea-685b09d5e228",
   "metadata": {},
   "source": [
    "## Reading the reference genome\n",
    "\n",
    "For this purpose we will use the pysam library.\n",
    "\n",
    "We will use the reference genome downloaded from http://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz.\n",
    "The reference genome was unzipped and indexed with `samtools faidx`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2137b685-222e-44ba-bad1-98d9d58aedab",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysam import FastaFile\n",
    "\n",
    "fasta_file = \"/home/priesgo/gencode/Homo_sapiens.GRCh38.dna.primary_assembly.fa\"\n",
    "reference_genome = FastaFile(fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a94db95-eff2-440e-911e-fb06a5667d33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCTGGTGCT'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fetch 9 bp between positions 20000 and 20009\n",
    "random_sequence = reference_genome.fetch(\"1\", 20000, 20009)\n",
    "random_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef229b5e-436f-4827-a5ef-9da4459b4403",
   "metadata": {},
   "source": [
    "### Manipulating sequences\n",
    "\n",
    "For this purpose the biopython library provides all that is required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d40b8dc-06f5-4bf7-95b1-294676c8d10a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('PGA', ExtendedIUPACProtein())"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "# translate DNA into protein sequence\n",
    "Seq(random_sequence).translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6d940f3c-c90e-4851-8fb6-2c1cb7efb316",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AGCACCAGG')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reverse complement sequences (handy for genes in the negative strand)\n",
    "Seq(random_sequence).reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "57435b21-7e2c-4d91-b540-c01a0c0f5a63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TFLR', ExtendedIUPACProtein())"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# you don't have to worry about casing\n",
    "Seq(\"acgTTTTTacgttt\").translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5e6bb98-aff7-4d5a-b60d-cecf0e7d06e8",
   "metadata": {},
   "source": [
    "## Reading the gene annotations\n",
    "\n",
    "We will rely on the library gtfparse which parses GTF files.\n",
    "\n",
    "We will use the GENCODE annotations downloaded from https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/gencode.v39.annotation.gtf.gz. This correspond to reference genome GRCh38."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b9b1e36b-6c55-42cb-8275-f140ec022f1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "import gtfparse\n",
    "\n",
    "gtf_file = \"/home/priesgo/gencode/gencode.v39.annotation.gtf.gz\"\n",
    "gtf = gtfparse.read_gtf(gtf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d7f5c3a4-b771-4cb6-ba9d-9fc945a76b95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end  score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1    chr1  HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2    chr1  HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3    chr1  HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4    chr1  HAVANA        exon  13221  14409    NaN      +     0   \n",
       "\n",
       "             gene_id                           gene_type  ...  \\\n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "1  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "2  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "3  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "4  ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "\n",
       "        transcript_type transcript_name transcript_support_level    tag  \\\n",
       "0                                                                         \n",
       "1  processed_transcript     DDX11L1-202                        1  basic   \n",
       "2  processed_transcript     DDX11L1-202                        1  basic   \n",
       "3  processed_transcript     DDX11L1-202                        1  basic   \n",
       "4  processed_transcript     DDX11L1-202                        1  basic   \n",
       "\n",
       "      havana_transcript exon_number            exon_id ont protein_id ccdsid  \n",
       "0                                                                             \n",
       "1  OTTHUMT00000362751.1                                                       \n",
       "2  OTTHUMT00000362751.1           1  ENSE00002234944.1                        \n",
       "3  OTTHUMT00000362751.1           2  ENSE00003582793.1                        \n",
       "4  OTTHUMT00000362751.1           3  ENSE00002312635.1                        \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3a8bd43b-ded0-4d79-b7fe-05aaa14aa8a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['transcribed_unprocessed_pseudogene', 'unprocessed_pseudogene',\n",
       "       'miRNA', 'lncRNA', 'protein_coding', 'processed_pseudogene',\n",
       "       'snRNA', 'transcribed_processed_pseudogene', 'misc_RNA', 'TEC',\n",
       "       'transcribed_unitary_pseudogene', 'snoRNA', 'scaRNA',\n",
       "       'rRNA_pseudogene', 'unitary_pseudogene', 'polymorphic_pseudogene',\n",
       "       'pseudogene', 'rRNA', 'IG_V_pseudogene', 'scRNA', 'IG_V_gene',\n",
       "       'IG_C_gene', 'IG_J_gene', 'sRNA', 'ribozyme',\n",
       "       'translated_processed_pseudogene', 'vault_RNA', 'TR_C_gene',\n",
       "       'TR_J_gene', 'TR_V_gene', 'TR_V_pseudogene',\n",
       "       'translated_unprocessed_pseudogene', 'TR_D_gene',\n",
       "       'IG_C_pseudogene', 'TR_J_pseudogene', 'IG_J_pseudogene',\n",
       "       'IG_D_gene', 'IG_pseudogene', 'Mt_tRNA', 'Mt_rRNA'], dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.gene_type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "92505464-3038-402e-b834-047ad6a2d215",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['', 'basic', 'basic,Ensembl_canonical', ...,\n",
       "       'basic,Ensembl_canonical,appris_principal_1,PAR',\n",
       "       'non_canonical_TEC,PAR',\n",
       "       'dotter_confirmed,non_canonical_TEC,not_best_in_genome_evidence,PAR'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.tag.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7e24ff00-06fe-434d-ad20-7bf04b40b5dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td></td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>OTTHUMT00000003223.4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td></td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>OTTHUMT00000003223.4</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td></td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>OTTHUMT00000003223.4</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td></td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>OTTHUMT00000003223.4</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>65565</td>\n",
       "      <td>65567</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td></td>\n",
       "      <td>RNA_Seq_supported_partial,basic,Ensembl_canoni...</td>\n",
       "      <td>OTTHUMT00000003223.4</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000493376.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  source      feature  start    end  score strand frame  \\\n",
       "58    chr1  HAVANA   transcript  65419  71585    NaN      +     0   \n",
       "59    chr1  HAVANA         exon  65419  65433    NaN      +     0   \n",
       "60    chr1  HAVANA         exon  65520  65573    NaN      +     0   \n",
       "61    chr1  HAVANA          CDS  65565  65573    NaN      +     0   \n",
       "62    chr1  HAVANA  start_codon  65565  65567    NaN      +     0   \n",
       "\n",
       "              gene_id       gene_type  ... transcript_type transcript_name  \\\n",
       "58  ENSG00000186092.7  protein_coding  ...  protein_coding       OR4F5-201   \n",
       "59  ENSG00000186092.7  protein_coding  ...  protein_coding       OR4F5-201   \n",
       "60  ENSG00000186092.7  protein_coding  ...  protein_coding       OR4F5-201   \n",
       "61  ENSG00000186092.7  protein_coding  ...  protein_coding       OR4F5-201   \n",
       "62  ENSG00000186092.7  protein_coding  ...  protein_coding       OR4F5-201   \n",
       "\n",
       "   transcript_support_level  \\\n",
       "58                            \n",
       "59                            \n",
       "60                            \n",
       "61                            \n",
       "62                            \n",
       "\n",
       "                                                  tag     havana_transcript  \\\n",
       "58  RNA_Seq_supported_partial,basic,Ensembl_canoni...  OTTHUMT00000003223.4   \n",
       "59  RNA_Seq_supported_partial,basic,Ensembl_canoni...  OTTHUMT00000003223.4   \n",
       "60  RNA_Seq_supported_partial,basic,Ensembl_canoni...  OTTHUMT00000003223.4   \n",
       "61  RNA_Seq_supported_partial,basic,Ensembl_canoni...  OTTHUMT00000003223.4   \n",
       "62  RNA_Seq_supported_partial,basic,Ensembl_canoni...  OTTHUMT00000003223.4   \n",
       "\n",
       "   exon_number            exon_id ont         protein_id ccdsid  \n",
       "58                                     ENSP00000493376.2         \n",
       "59           1  ENSE00003812156.1      ENSP00000493376.2         \n",
       "60           2  ENSE00003813641.1      ENSP00000493376.2         \n",
       "61           2  ENSE00003813641.1      ENSP00000493376.2         \n",
       "62           2  ENSE00003813641.1      ENSP00000493376.2         \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter to protein coding transcripts tagged as basic\n",
    "filtered_gtf = gtf[(gtf.gene_type == \"protein_coding\") & (gtf.tag.str.contains(\"basic\"))]\n",
    "filtered_gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d85dfb95-7dda-487a-803b-8f5db8738b06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69037</td>\n",
       "      <td>71585</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003813949.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  start    end      transcript_id            exon_id\n",
       "59    chr1  65419  65433  ENST00000641515.2  ENSE00003812156.1\n",
       "60    chr1  65520  65573  ENST00000641515.2  ENSE00003813641.1\n",
       "63    chr1  69037  71585  ENST00000641515.2  ENSE00003813949.1"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the list of exons for the first transcript\n",
    "for transcript in filtered_gtf.transcript_id:\n",
    "    exons = filtered_gtf[(filtered_gtf.transcript_id == transcript) & (filtered_gtf.feature == \"exon\")][[\"seqname\", \"start\", \"end\", \"transcript_id\", \"exon_id\"]]\n",
    "    break\n",
    "    \n",
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "aa09f89d-5e10-45d2-96ca-48f411ef26a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('PDLFSFMPHSVKIAVVSSSYEEVTAEAISWNESTSETNNSMVTEFIFLGLSDSQ...FKK', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for a given list of exons, fetch the corresponding sequence from the reference genome, join into the CDS and translate into protein sequence\n",
    "cds = \"\"\n",
    "for _, exon in exons.iterrows():\n",
    "    cds = cds + reference_genome.fetch(exon.seqname.strip(\"chr\"), exon.start, exon.end)\n",
    "        \n",
    "Seq(cds).translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c4f754d-166c-4c29-9e9d-b1a705bb9471",
   "metadata": {},
   "source": [
    "## Reading mutations from a VCF file\n",
    "\n",
    "We will rely on the cyvcf2 which is able to read and write VCF files.\n",
    "\n",
    "For testing purposes we will use the ClinVar VCF file in GRCh38 downloaded from https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "391d3c51-fb0e-4852-b3fb-2dcfdf52b5ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::hts_idx_load3] The index file is older than the data file: /home/priesgo/gencode/clinvar.vcf.gz.tbi\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NW_009646201.1\t83614\t17735\tTC\tT\t.\t.\tALLELEID=32774;CLNDISDB=Human_Phenotype_Ontology:HP:0032224,MedGen:C0000778,OMIM:616093,SNOMED_CT:63915006|MedGen:CN517202;CLNDN=ABO_blood_group_system|not_provided;CLNHGVS=NW_009646201.1:g.83615del;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Benign;CLNVC=Deletion;CLNVCSO=SO:0000159;CLNVI=OMIM_Allelic_Variant:110300.0001;GENEINFO=ABO:28;ORIGIN=1;RS=1556058284\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from cyvcf2 import VCF\n",
    "\n",
    "vcf_file = gtf_file = \"/home/priesgo/gencode/clinvar.vcf.gz\"\n",
    "vcf = VCF(vcf_file)\n",
    "for v in vcf:\n",
    "    # filter variants where FILTER != PASS\n",
    "    if v.FILTER is None:\n",
    "        variant = v\n",
    "print(variant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "20face0a-dc3d-4177-87c7-d85004b84263",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant.FILTER is None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e930599b-6240-4ec4-b31e-9073810045ac",
   "metadata": {},
   "source": [
    "## Putting it all together"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c0cb710-3664-46e7-aa12-69186ed20216",
   "metadata": {},
   "source": [
    "### Fetch the transcripts overlapped by a mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "bd741267-3158-4436-a574-2099d4b8d12c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-213</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,basic,appris_principal_5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000480678.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1020</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>925731</td>\n",
       "      <td>944574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,CCDS</td>\n",
       "      <td>OTTHUMT00000276866.3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000342313.3</td>\n",
       "      <td>CCDS2.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqname  source     feature   start     end  score strand frame  \\\n",
       "941     chr1  HAVANA  transcript  923923  944574    NaN      +     0   \n",
       "974     chr1  HAVANA  transcript  923923  944574    NaN      +     0   \n",
       "1020    chr1  HAVANA  transcript  925731  944574    NaN      +     0   \n",
       "\n",
       "                 gene_id       gene_type  ... transcript_type transcript_name  \\\n",
       "941   ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "974   ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-213   \n",
       "1020  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-202   \n",
       "\n",
       "     transcript_support_level  \\\n",
       "941                         5   \n",
       "974                         5   \n",
       "1020                        5   \n",
       "\n",
       "                                                    tag     havana_transcript  \\\n",
       "941   CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "974         CAGE_supported_TSS,basic,appris_principal_5                         \n",
       "1020                                         basic,CCDS  OTTHUMT00000276866.3   \n",
       "\n",
       "     exon_number exon_id ont         protein_id   ccdsid  \n",
       "941                           ENSP00000478421.2           \n",
       "974                           ENSP00000480678.2           \n",
       "1020                          ENSP00000342313.3  CCDS2.2  \n",
       "\n",
       "[3 rows x 25 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE: unfortunately chromosomes in Clinvar are missing the chr in front of them\n",
    "# there is a resource for mapping chromosomes between nomenclatures, we may need it at some point...\n",
    "overlapping_transcripts = filtered_gtf[(filtered_gtf.seqname == \"chr\" + variant.CHROM) & (filtered_gtf.start <= variant.POS) & (filtered_gtf.end > variant.POS) & (filtered_gtf.feature == \"transcript\")]\n",
    "overlapping_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "197a634e-a2e3-4b01-9c5a-31e5a1a169a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432</td>\n",
       "      <td>924948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001637883.3</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>925922</td>\n",
       "      <td>926013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003794726.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>930155</td>\n",
       "      <td>930336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003915575.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>931039</td>\n",
       "      <td>931089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003911839.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>935772</td>\n",
       "      <td>935896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003911340.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>939040</td>\n",
       "      <td>939129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003912201.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>939275</td>\n",
       "      <td>939412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSE00001708361.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>941144</td>\n",
       "      <td>941306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00003477353.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>942136</td>\n",
       "      <td>942251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSE00003913392.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>942410</td>\n",
       "      <td>942488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSE00003675531.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>942559</td>\n",
       "      <td>943058</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSE00002728091.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>966</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>943253</td>\n",
       "      <td>943377</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSE00002733131.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>968</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>943698</td>\n",
       "      <td>943808</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSE00002692620.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>943908</td>\n",
       "      <td>944150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSE00003845896.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname  source feature   start     end  score strand frame  \\\n",
       "943    chr1  HAVANA     CDS  924432  924948    NaN      +     0   \n",
       "946    chr1  HAVANA     CDS  925922  926013    NaN      +     2   \n",
       "948    chr1  HAVANA     CDS  930155  930336    NaN      +     0   \n",
       "950    chr1  HAVANA     CDS  931039  931089    NaN      +     1   \n",
       "952    chr1  HAVANA     CDS  935772  935896    NaN      +     1   \n",
       "954    chr1  HAVANA     CDS  939040  939129    NaN      +     2   \n",
       "956    chr1  HAVANA     CDS  939275  939412    NaN      +     2   \n",
       "958    chr1  HAVANA     CDS  941144  941306    NaN      +     2   \n",
       "960    chr1  HAVANA     CDS  942136  942251    NaN      +     1   \n",
       "962    chr1  HAVANA     CDS  942410  942488    NaN      +     2   \n",
       "964    chr1  HAVANA     CDS  942559  943058    NaN      +     1   \n",
       "966    chr1  HAVANA     CDS  943253  943377    NaN      +     2   \n",
       "968    chr1  HAVANA     CDS  943698  943808    NaN      +     0   \n",
       "970    chr1  HAVANA     CDS  943908  944150    NaN      +     0   \n",
       "\n",
       "                gene_id       gene_type  ... transcript_type transcript_name  \\\n",
       "943  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "946  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "948  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "950  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "952  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "954  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "956  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "958  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "960  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "962  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "964  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "966  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "968  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "970  ENSG00000187634.13  protein_coding  ...  protein_coding      SAMD11-209   \n",
       "\n",
       "    transcript_support_level  \\\n",
       "943                        5   \n",
       "946                        5   \n",
       "948                        5   \n",
       "950                        5   \n",
       "952                        5   \n",
       "954                        5   \n",
       "956                        5   \n",
       "958                        5   \n",
       "960                        5   \n",
       "962                        5   \n",
       "964                        5   \n",
       "966                        5   \n",
       "968                        5   \n",
       "970                        5   \n",
       "\n",
       "                                                   tag     havana_transcript  \\\n",
       "943  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "946  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "948  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "950  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "952  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "954  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "956  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "958  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "960  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "962  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "964  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "966  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "968  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "970  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "\n",
       "    exon_number            exon_id ont         protein_id ccdsid  \n",
       "943           1  ENSE00001637883.3      ENSP00000478421.2         \n",
       "946           2  ENSE00003794726.1      ENSP00000478421.2         \n",
       "948           3  ENSE00003915575.1      ENSP00000478421.2         \n",
       "950           4  ENSE00003911839.1      ENSP00000478421.2         \n",
       "952           5  ENSE00003911340.1      ENSP00000478421.2         \n",
       "954           6  ENSE00003912201.1      ENSP00000478421.2         \n",
       "956           7  ENSE00001708361.1      ENSP00000478421.2         \n",
       "958           8  ENSE00003477353.1      ENSP00000478421.2         \n",
       "960           9  ENSE00003913392.1      ENSP00000478421.2         \n",
       "962          10  ENSE00003675531.1      ENSP00000478421.2         \n",
       "964          11  ENSE00002728091.1      ENSP00000478421.2         \n",
       "966          12  ENSE00002733131.1      ENSP00000478421.2         \n",
       "968          13  ENSE00002692620.1      ENSP00000478421.2         \n",
       "970          14  ENSE00003845896.1      ENSP00000478421.2         \n",
       "\n",
       "[14 rows x 25 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fetch all exons for a given transcript\n",
    "# NOTE: beware to read elements of feature_type=CDS and not feature_type=exon, the later include the UTR regions which we do not want\n",
    "transcript_id = overlapping_transcripts.transcript_id.iloc[0]\n",
    "exons = filtered_gtf[(filtered_gtf.transcript_id == transcript_id) & (filtered_gtf.feature == \"CDS\")].sort_values(\"start\")\n",
    "exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9997c129-b1ad-4bad-b60c-871ae858fd52",
   "metadata": {},
   "source": [
    "### Fetch the genomic sequence for each of the exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "5794d6fd-f3d4-4094-9888-561801d112c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sequence_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924432</td>\n",
       "      <td>924948</td>\n",
       "      <td>ATGCCGGCGGTCAAGAAGGAGTTCCCGGGCCGCGAGGACCTGGCCC...</td>\n",
       "      <td>517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925922</td>\n",
       "      <td>926013</td>\n",
       "      <td>GGAAAAGTCTGAAGACGCTTATGTCCAAGGGGATCCTGCAGGTGCA...</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>chr1</td>\n",
       "      <td>930155</td>\n",
       "      <td>930336</td>\n",
       "      <td>AACCGGGGGCGGCTGGCAGACAAGAGGACAGTCGCCCTGCCTGCCG...</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>chr1</td>\n",
       "      <td>931039</td>\n",
       "      <td>931089</td>\n",
       "      <td>AGTCCACACCCACTGGGACGTGAACATCTCTTTCCGAGAGGCGTCC...</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>chr1</td>\n",
       "      <td>935772</td>\n",
       "      <td>935896</td>\n",
       "      <td>CCAGGACGGCAACCTTCCCACCCTCATATCCAGCGTCCACCGCAGC...</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>chr1</td>\n",
       "      <td>939040</td>\n",
       "      <td>939129</td>\n",
       "      <td>GTGACCTGTTGGGCAAGAGGCTGGGCCGCTCCCCCCGTATCAGCAG...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>chr1</td>\n",
       "      <td>939275</td>\n",
       "      <td>939412</td>\n",
       "      <td>AGGCGCTGCTGCTGCCGCGGGAGCTGGGGCCCAGCATGGCCCCGGA...</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>chr1</td>\n",
       "      <td>941144</td>\n",
       "      <td>941306</td>\n",
       "      <td>ATCTCCTGAGGGTCCGGCAGGAGGTGGCGGCTGCAGCTCTGAGGGG...</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942136</td>\n",
       "      <td>942251</td>\n",
       "      <td>GGAGCTGCCTCAGCCGCCCCCCTTGCTGTCGCCGCAGAATGCCCCT...</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942410</td>\n",
       "      <td>942488</td>\n",
       "      <td>GCTACGGCTTCCTGCCCCCCGCGCAGGCGGAGATGTTCGCCTGGCA...</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942559</td>\n",
       "      <td>943058</td>\n",
       "      <td>GCTGGAGCTGCCCGCCGACCTCCTGCGGCAGAAGGAGCTGGAGAGC...</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>966</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943253</td>\n",
       "      <td>943377</td>\n",
       "      <td>GCGCGGTAGGGGGACTCTCCATGGATGGGGAGGAGGCCCCAGCCCC...</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>968</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943698</td>\n",
       "      <td>943808</td>\n",
       "      <td>GTCTTCAGGGAGCAGGGGATCGACGGGGAGACCCTGCCACTGCTGA...</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943908</td>\n",
       "      <td>944150</td>\n",
       "      <td>GTGGCCAGGCGCCTGGGCCGAGTTTTCTACGTGGCCAGCTTCCCCG...</td>\n",
       "      <td>243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname   start     end  \\\n",
       "943    chr1  924432  924948   \n",
       "946    chr1  925922  926013   \n",
       "948    chr1  930155  930336   \n",
       "950    chr1  931039  931089   \n",
       "952    chr1  935772  935896   \n",
       "954    chr1  939040  939129   \n",
       "956    chr1  939275  939412   \n",
       "958    chr1  941144  941306   \n",
       "960    chr1  942136  942251   \n",
       "962    chr1  942410  942488   \n",
       "964    chr1  942559  943058   \n",
       "966    chr1  943253  943377   \n",
       "968    chr1  943698  943808   \n",
       "970    chr1  943908  944150   \n",
       "\n",
       "                                              sequence  sequence_len  \n",
       "943  ATGCCGGCGGTCAAGAAGGAGTTCCCGGGCCGCGAGGACCTGGCCC...           517  \n",
       "946  GGAAAAGTCTGAAGACGCTTATGTCCAAGGGGATCCTGCAGGTGCA...            92  \n",
       "948  AACCGGGGGCGGCTGGCAGACAAGAGGACAGTCGCCCTGCCTGCCG...           182  \n",
       "950  AGTCCACACCCACTGGGACGTGAACATCTCTTTCCGAGAGGCGTCC...            51  \n",
       "952  CCAGGACGGCAACCTTCCCACCCTCATATCCAGCGTCCACCGCAGC...           125  \n",
       "954  GTGACCTGTTGGGCAAGAGGCTGGGCCGCTCCCCCCGTATCAGCAG...            90  \n",
       "956  AGGCGCTGCTGCTGCCGCGGGAGCTGGGGCCCAGCATGGCCCCGGA...           138  \n",
       "958  ATCTCCTGAGGGTCCGGCAGGAGGTGGCGGCTGCAGCTCTGAGGGG...           163  \n",
       "960  GGAGCTGCCTCAGCCGCCCCCCTTGCTGTCGCCGCAGAATGCCCCT...           116  \n",
       "962  GCTACGGCTTCCTGCCCCCCGCGCAGGCGGAGATGTTCGCCTGGCA...            79  \n",
       "964  GCTGGAGCTGCCCGCCGACCTCCTGCGGCAGAAGGAGCTGGAGAGC...           500  \n",
       "966  GCGCGGTAGGGGGACTCTCCATGGATGGGGAGGAGGCCCCAGCCCC...           125  \n",
       "968  GTCTTCAGGGAGCAGGGGATCGACGGGGAGACCCTGCCACTGCTGA...           111  \n",
       "970  GTGGCCAGGCGCCTGGGCCGAGTTTTCTACGTGGCCAGCTTCCCCG...           243  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_sequences = []\n",
    "for _, exon in exons.iterrows():\n",
    "    # NOTE: beware that coordinates in the GTF are 1-based but when querying the genome when pysam we need 0-based\n",
    "    exon_sequences.append(reference_genome.fetch(exon.seqname.strip(\"chr\"), exon.start - 1, exon.end))\n",
    "\n",
    "exons[\"sequence\"] = exon_sequences\n",
    "exons[\"sequence_len\"] = exons[\"sequence\"].transform(len)\n",
    "exons[[\"seqname\", \"start\", \"end\", \"sequence\", \"sequence_len\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7d95a09-7f2e-44be-b445-9bde861bf498",
   "metadata": {},
   "source": [
    "### Apply the mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "cc8a54cd-c1f8-4dae-a141-9ab3b2908d49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sequence_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>925922</td>\n",
       "      <td>926013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>CAGE_supported_TSS,RNA_Seq_supported_partial,b...</td>\n",
       "      <td>OTTHUMT00000316521.3</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003794726.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000478421.2</td>\n",
       "      <td></td>\n",
       "      <td>GGAAAAGTCTGAAGACGCTTATGTCCAAGGGGATCCTGCAGGTGCA...</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname  source feature   start     end  score strand frame  \\\n",
       "946    chr1  HAVANA     CDS  925922  926013    NaN      +     2   \n",
       "\n",
       "                gene_id       gene_type  ... transcript_support_level  \\\n",
       "946  ENSG00000187634.13  protein_coding  ...                        5   \n",
       "\n",
       "                                                   tag     havana_transcript  \\\n",
       "946  CAGE_supported_TSS,RNA_Seq_supported_partial,b...  OTTHUMT00000316521.3   \n",
       "\n",
       "    exon_number            exon_id ont         protein_id ccdsid  \\\n",
       "946           2  ENSE00003794726.1      ENSP00000478421.2          \n",
       "\n",
       "                                              sequence sequence_len  \n",
       "946  GGAAAAGTCTGAAGACGCTTATGTCCAAGGGGATCCTGCAGGTGCA...           92  \n",
       "\n",
       "[1 rows x 27 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# identify the exon overlapped by the mutation\n",
    "overlapping_exon = exons[(exons.seqname.str.strip(\"chr\") == variant.CHROM) & (exons.start <= variant.POS) & (exons.end > variant.POS)]\n",
    "overlapping_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "0944b0bb-8ce2-4ea4-b617-8d90e7ad323e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GGAAAAGTCTGAAGACGCTTATGTCCAAGGGGATCCTGCAGGTGCATCCTCCGATCTGCGACTGCCCGGGCTGCCGAATATCCTCCCCGGTG'"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlapping_sequence = overlapping_exon.sequence.iloc[0]\n",
    "overlapping_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "74563820-2a23-40f7-be7f-f98d0f46e901",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GGAAAAGTCTGAAGACGCTTATGTCCAAGGaGATCCTGCAGGTGCATCCTCCGATCTGCGACTGCCCGGGCTGCCGAATATCCTCCCCGGTG'"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# obtains the mutated sequence\n",
    "variant_relative_position = variant.POS - overlapping_exon.start.iloc[0]\n",
    "print(variant_relative_position)\n",
    "mutated_sequence = overlapping_sequence[0: variant_relative_position] + variant.ALT[0].lower() + overlapping_sequence[variant_relative_position + 1 : ]\n",
    "mutated_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0360466c-b0af-473e-a3c6-1f5a0ae871c2",
   "metadata": {},
   "source": [
    "### Virtual splicing and translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "2d2e8e5a-d65e-4ff6-bf09-7c9861d9c20d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGCCGGCGGTCAAGAAGGAGTTCCCGGGCCGCGAGGACCTGGCCCTGGCTCTGGCCACGTTCCACCCGACCCTGGCCGCGCTGCCGCTGCCGCCGCTGCCAGGCTACCTGGCGCCACTGCCCGCGGCGGCCGCCCTCCCCCCGGCCGCCTCGCTGCCCGCCTCGGCCGCCGGTTACGAGGCTCTGCTGGCCCCGCCGCTCCGCCCCCCGCGCGCCTACCTCAGCCTGCACGAGGCCGCCCCGCACCTCCACCTGCCCAGGGACCCGCTGGCCCTCGAGCGCTTCTCGGCCACCGCGGCCGCGGCCCCGGATTTCCAGCCGCTGCTGGACAACGGCGAGCCGTGCATCGAGGTGGAGTGCGGCGCCAACCGCGCGCTGCTCTACGTGCGCAAACTCTGCCAGGGCAGCAAGGGCCCGTCCATCCGCCACCGCGGCGAGTGGCTCACGCCCAACGAGTTCCAGTTCGTCAGCGGCCGCGAGACGGCCAAGGACTGGAAGCGCAGCATCCGCCACAAAGGGAAAAGTCTGAAGACGCTTATGTCCAAGGGGATCCTGCAGGTGCATCCTCCGATCTGCGACTGCCCGGGCTGCCGAATATCCTCCCCGGTGAACCGGGGGCGGCTGGCAGACAAGAGGACAGTCGCCCTGCCTGCCGCCCGGAACCTGAAGAAGGAGCGAACTCCCAGCTTCTCTGCCAGCGATGGTGACAGCGACGGGAGTGGCCCCACCTGTGGGCGGCGGCCAGGCTTGAAGCAGGAGGATGGTCCGCACATCCGTATCATGAAGAGAAGAGTCCACACCCACTGGGACGTGAACATCTCTTTCCGAGAGGCGTCCTGCAGCCAGGACGGCAACCTTCCCACCCTCATATCCAGCGTCCACCGCAGCCGCCACCTCGTTATGCCCGAGCATCAGAGCCGCTGTGAATTCCAGAGAGGCAGCCTGGAGATTGGCCTGCGACCCGCCGGTGACCTGTTGGGCAAGAGGCTGGGCCGCTCCCCCCGTATCAGCAGCGACTGCTTTTCAGAGAAGAGGGCACGAAGCGAATCGCCTCAAGAGGCGCTGCTGCTGCCGCGGGAGCTGGGGCCCAGCATGGCCCCGGAGGACCATTACCGCCGGCTTGTGTCAGCACTGAGCGAGGCCAGCACCTTTGAGGACCCTCAGCGCCTCTACCACCTGGGCCTCCCCAGCCACGATCTCCTGAGGGTCCGGCAGGAGGTGGCGGCTGCAGCTCTGAGGGGCCCCAGTGGCCTGGAAGCCCACCTGCCCTCCTCCACGGCAGGTCAGCGTCGGAAGCAGGGCCTGGCTCAGCACCGGGAGGGCGCCGCCCCAGCTGCCGCCCCGTCCTTCTCGGAGAGGGAGCTGCCTCAGCCGCCCCCCTTGCTGTCGCCGCAGAATGCCCCTCACGTCGCCCTGGGCCCCCATCTCAGGCCCCCCTTCCTGGGGGTGCCCTCGGCTCTGTGCCAGACCCCAGGCTACGGCTTCCTGCCCCCCGCGCAGGCGGAGATGTTCGCCTGGCAGCAGGAGCTCCTGCGGAAGCAGAACCTGGCCCGGCTGGAGCTGCCCGCCGACCTCCTGCGGCAGAAGGAGCTGGAGAGCGCGCGCCCACAGCTGCTGGCGCCCGAGACCGCCCTGCGCCCCAACGACGGCGCCGAGGAGCTGCAGCGGCGCGGGGCCCTGCTGGTGCTGAACCACGGCGCGGCGCCACTGCTGGCCCTGCCCCCCCAGGGGCCCCCGGGCTCCGGACCCCCCACCCCGTCCCGGGACTCTGCCCGGCGAGCCCCCCGGAAGGGGGGTCCCGGCCCTGCCTCAGCGCGGCCCAGCGAGTCCAAGGAGATGACGGGGGCTAGGCTCTGGGCACAAGATGGCTCGGAAGACGAGCCCCCCAAAGACTCGGACGGAGAGGACCCCGAGACGGCAGCTGTTGGGTGCAGGGGGCCCACTCCGGGCCAAGCTCCAGCTGGAGGGGCCGGCGCCGAGGGGAAGGGGCTTTTCCCAGGGTCCACACTGCCCCTGGGCTTCCCTTATGCCGTCAGCCCCTACTTCCACACAGGCGCGGTAGGGGGACTCTCCATGGATGGGGAGGAGGCCCCAGCCCCTGAGGACGTCACCAAGTGGACCGTGGATGACGTCTGCAGCTTCGTGGGGGGCCTGTCTGGCTGTGGAGAGTACACTCGGGTCTTCAGGGAGCAGGGGATCGACGGGGAGACCCTGCCACTGCTGACGGAGGAGCACCTGCTGACCAACATGGGGCTGAAGCTGGGGCCCGCCCTCAAGATCCGGGCCCAGGTGGCCAGGCGCCTGGGCCGAGTTTTCTACGTGGCCAGCTTCCCCGTGGCTCTGCCACTGCAGCCACCAACCCTGCGGGCCCCGGAGCGAGAACTCGGCACAGGAGAGCAGCCCTTGTCCCCCACGACGGCCACGTCCCCCTATGGAGGGGGCCACGCCCTTGCCGGTCAAACTTCACCCAAGCAGGAGAATGGGACCTTGGCTCTACTTCCAGGGGCCCCCGACCCTTCCCAGCCTCTGTGT'"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_sequence = \"\"\n",
    "for _, exon in exons.iterrows():\n",
    "    dna_sequence = dna_sequence + exon.sequence\n",
    "    \n",
    "dna_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "2e1f0eab-d13f-405c-8b16-65dc9b78af29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAALPPAASLPASAAGYEALLAPPLRPPRAYLSLHEAAPHLHLPRDPLALERFSATAAAAPDFQPLLDNGEPCIEVECGANRALLYVRKLCQGSKGPSIRHRGEWLTPNEFQFVSGRETAKDWKRSIRHKGKSLKTLMSKGILQVHPPICDCPGCRISSPVNRGRLADKRTVALPAARNLKKERTPSFSASDGDSDGSGPTCGRRPGLKQEDGPHIRIMKRRVHTHWDVNISFREASCSQDGNLPTLISSVHRSRHLVMPEHQSRCEFQRGSLEIGLRPAGDLLGKRLGRSPRISSDCFSEKRARSESPQEALLLPRELGPSMAPEDHYRRLVSALSEASTFEDPQRLYHLGLPSHDLLRVRQEVAAAALRGPSGLEAHLPSSTAGQRRKQGLAQHREGAAPAAAPSFSERELPQPPPLLSPQNAPHVALGPHLRPPFLGVPSALCQTPGYGFLPPAQAEMFAWQQELLRKQNLARLELPADLLRQKELESARPQLLAPETALRPNDGAEELQRRGALLVLNHGAAPLLALPPQGPPGSGPPTPSRDSARRAPRKGGPGPASARPSESKEMTGARLWAQDGSEDEPPKDSDGEDPETAAVGCRGPTPGQAPAGGAGAEGKGLFPGSTLPLGFPYAVSPYFHTGAVGGLSMDGEEAPAPEDVTKWTVDDVCSFVGGLSGCGEYTRVFREQGIDGETLPLLTEEHLLTNMGLKLGPALKIRAQVARRLGRVFYVASFPVALPLQPPTLRAPERELGTGEQPLSPTTATSPYGGGHALAGQTSPKQENGTLALLPGAPDPSQPLC'"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_protein = str(Seq(dna_sequence).translate())\n",
    "original_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "7982ef7a-8cbd-450b-b88a-4cf5f2fc8267",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAALPPAASLPASAAGYEALLAPPLRPPRAYLSLHEAAPHLHLPRDPLALERFSATAAAAPDFQPLLDNGEPCIEVECGANRALLYVRKLCQGSKGPSIRHRGEWLTPNEFQFVSGRETAKDWKRSIRHKGKSLKTLMSKEILQVHPPICDCPGCRISSPVNRGRLADKRTVALPAARNLKKERTPSFSASDGDSDGSGPTCGRRPGLKQEDGPHIRIMKRRVHTHWDVNISFREASCSQDGNLPTLISSVHRSRHLVMPEHQSRCEFQRGSLEIGLRPAGDLLGKRLGRSPRISSDCFSEKRARSESPQEALLLPRELGPSMAPEDHYRRLVSALSEASTFEDPQRLYHLGLPSHDLLRVRQEVAAAALRGPSGLEAHLPSSTAGQRRKQGLAQHREGAAPAAAPSFSERELPQPPPLLSPQNAPHVALGPHLRPPFLGVPSALCQTPGYGFLPPAQAEMFAWQQELLRKQNLARLELPADLLRQKELESARPQLLAPETALRPNDGAEELQRRGALLVLNHGAAPLLALPPQGPPGSGPPTPSRDSARRAPRKGGPGPASARPSESKEMTGARLWAQDGSEDEPPKDSDGEDPETAAVGCRGPTPGQAPAGGAGAEGKGLFPGSTLPLGFPYAVSPYFHTGAVGGLSMDGEEAPAPEDVTKWTVDDVCSFVGGLSGCGEYTRVFREQGIDGETLPLLTEEHLLTNMGLKLGPALKIRAQVARRLGRVFYVASFPVALPLQPPTLRAPERELGTGEQPLSPTTATSPYGGGHALAGQTSPKQENGTLALLPGAPDPSQPLC'"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_dna_sequence = \"\"\n",
    "for _, exon in exons.iterrows():\n",
    "    if exon.exon_id == overlapping_exon.exon_id.iloc[0]:\n",
    "        mutated_dna_sequence = mutated_dna_sequence + mutated_sequence\n",
    "    else:\n",
    "        mutated_dna_sequence = mutated_dna_sequence + exon.sequence\n",
    "    \n",
    "mutated_protein = str(Seq(mutated_dna_sequence).translate())\n",
    "mutated_protein"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fe59b59-529a-4376-a6d4-042d39979792",
   "metadata": {},
   "source": [
    "### Find the mutation and build the neoantigen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "56ce30ea-d44d-435e-a924-f59ea4fa93d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "182"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_position = 0\n",
    "for ref, alt in zip(original_protein, mutated_protein):\n",
    "    if ref != alt:\n",
    "        break\n",
    "    aa_position += 1\n",
    "    \n",
    "aa_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "e6136ed9-64f6-4dcb-9efc-231e5bfcceee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'RHKGKSLKTLMSKGILQVHPPICDCPG'"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_protein[aa_position - 13: aa_position + 14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "cb3e61c2-cb61-4532-9619-f23f1593afa6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'RHKGKSLKTLMSKEILQVHPPICDCPG'"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_protein[aa_position - 13: aa_position + 14]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
